### PR TITLE
feat(sink): handle stream key sink pk mismatch

### PR DESCRIPTION
**Pull request by st1page**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

fix https://github.com/risingwavelabs/risingwave/issues/12026
fix https://github.com/risingwavelabs/risingwave/issues/12447

the algorithm is https://github.com/risingwavelabs/risingwave/issues/12026#issuecomment-1704641675

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/common/src/types/interval.rs [hunk 1]
```diff
@@ -44,7 +44,7 @@ use crate::estimate_size::EstimateSize;
 /// This internals is learned from PG:
 /// <https://www.postgresql.org/docs/9.1/datatype-datetime.html#:~:text=field%20is%20negative.-,Internally,-interval%20values%20are>
 #[derive(Debug, Clone, Copy, Default, EstimateSize)]
-pub struct Interval {
+pub struct  Interval {
```
Comment (st1page) [inline src/common/src/types/interval.rs:47]:
```suggestion
pub struct Interval {
```

**Comment by st1page** [comment]
blocked on https://github.com/risingwavelabs/risingwave/issues/12460

**Comment by StrikeW** [comment]
Could you add some context for the problem that would be helpful for future reference?

File: src/stream/src/executor/sink.rs [hunk 2]
```diff
@@ -156,36 +179,80 @@ impl<F: LogStoreFactory> SinkExecutor<F> {
...
-                        force_append_only(chunk)
-                    } else {
-                        chunk
-                    };
-
-                    log_writer.write_chunk(chunk.clone()).await?;
-
-                    // Use original chunk instead of the reordered one as the executor output.
-                    yield Message::Chunk(chunk);
+        if stream_key_sink_pk_mismatch {
```
Comment (fuyufjh) [inline src/stream/src/executor/sink.rs:182]: Recommend adding some comments here because the solution is tricky to newcomers.

**Review by fuyufjh — APPROVED** [review]

**Review by hzxa21 — APPROVED** [review]

File: src/tests/simulation/tests/integration_tests/sink/basic.rs [hunk 3]
```diff
@@ -177,6 +177,7 @@ async fn test_sink_basic() -> Result<()> {
             sleep(Duration::from_millis(10)).await;
         }
     }
+    sleep(Duration::from_millis(10000)).await;
```
Comment (st1page) [inline src/tests/simulation/tests/integration_tests/sink/basic.rs:180]: c.c. @wenym1

**Pull request merged**